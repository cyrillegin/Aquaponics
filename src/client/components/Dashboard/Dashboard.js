import React, { useEffect, useState } from 'react';
import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import Graph from '../../charts/Graph';
import { windowEmitter, searchToObject } from '../../utilities/Window';
import SensorDetails from '../SensorDetails';

const Dashboard = ({ className, stations, dashboards }) => {
  const [currentStations, changeStations] = useState([]);
  const [currentSensor, changeSensor] = useState({});
  const [currentDashboard, changeDashboard] = useState({});

  useEffect(() => {
    const setStations = () => {
      const { station, sensor, dashboard } = searchToObject();
      if (station) {
        const stationId = parseInt(station, 10);
        const currentStation = stations.filter(sta => sta.id === stationId)[0];
        changeStations([currentStation]);
      } else if (sensor) {
        const sensorId = parseInt(sensor, 10);
        const currentStation = stations.find(sta => sta.sensors.find(sen => sen.id === sensorId));
        // console.log(currentStation);
        changeStations([currentStation]);
        const curSensor = currentStation.sensors.find(sen => sen.id === sensorId);
        changeSensor(curSensor);
      } else {
        changeDashboard(dashboards.find(dash => dash.dashboardId === dashboard));
      }
    };

    setStations();
    windowEmitter.listen('change', () => {
      setStations();
    });
  }, [stations]);

  return (
    <div className={className}>
      {currentSensor.id && (
        <div className="sensor tall">
          <Graph station={currentStations[0]} sensor={currentSensor} renderTrigger={new Date()} />
          <SensorDetails sensor={currentSensor} />
        </div>
      )}

      {!currentSensor.id && currentStations.length && (
        <>
          {currentStations.map(station => (
            <div className="station" key={station.id}>
              {station.sensors.map(sensor => (
                <div className="sensor" key={sensor.id}>
                  <Graph station={station.name} sensor={sensor} renderTrigger={new Date()} />
                </div>
              ))}
            </div>
          ))}
        </>
      )}

      {currentDashboard && currentDashboard.sensors && (
        <div className="station">
          {currentDashboard.sensors.map(sensor => (
            <div className="sensor" key={sensor.id}>
              <Graph name={currentDashboard.name} sensor={sensor} renderTrigger={new Date()} />
            </div>
          ))}
        </div>
      )}

      {!currentDashboard.sensors && !currentSensor.id && !currentStations.length && <>Loading...</>}
    </div>
  );
};

Dashboard.propTypes = {
  className: PropTypes.string.isRequired,
  stations: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      health: PropTypes.oneOf(['healthy', 'unhealthy']),
      sensors: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.number.isRequired,
          name: PropTypes.string.isRequired,
          health: PropTypes.oneOf(['healthy', 'unhealthy']),
        }),
      ),
    }),
  ).isRequired,
  dashboards: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      sensors: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.number.isRequired,
          stationId: PropTypes.number.isRequired,
          position: PropTypes.number.isRequired,
        }),
      ),
    }),
  ).isRequired,
};

const styledDashboard = styled(Dashboard)`
  position: absolute;
  left: 400px;
  background: white;
  width: calc(100% - 408px);
  margin-top: 5rem;
  margin-left: 8px;

  .station {
    width: 100%;
    background: white;
    display: grid;
    grid-template-columns: 1fr 1fr 1fr;
    margin: 2rem 0;
  }
  .sensor {
    width: 100%;
    height: 300px;

    &.tall {
      height: 600px;
    }
  }
`;

export default styledDashboard;
