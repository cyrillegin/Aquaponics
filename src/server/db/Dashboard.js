import { Model, DataTypes } from 'sequelize';

class Dashboard extends Model {}

function buildDashboardSchema(sequelize) {
  Dashboard.init(
    {
      name: { type: DataTypes.STRING, allowNull: false },
      dashboardId: { type: DataTypes.STRING, allowNull: false, field: 'dashboard_id' },
      sensorId: { type: DataTypes.INTEGER, allowNull: false, field: 'sensor_id' },
      station_Id: { type: DataTypes.INTEGER, allowNull: false, field: 'station_id' },
      position: { type: DataTypes.INTEGER, allowNull: false },
    },
    {
      sequelize,
      modelName: 'dashboard',
      createdAt: 'created_at',
      updatedAt: 'updated_at',
      paranoid: true,
      underscored: true,
    },
  );
}

export { buildDashboardSchema, Dashboard };
