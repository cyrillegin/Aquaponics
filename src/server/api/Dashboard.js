import { Router } from 'express';
import { Dashboard, DashboardGraph, DashboardSource } from '../db';
import { validateDashboardParams } from '../utilities/Validators';

const router = new Router();

/**
 * GET
 * Gets an array of dashboards
 *
 * returns {
 *   dashboards: [{
 *     id: 123,
 *     name: 'asdf',
 *     dashboardGraph: [{
 *       dashboardId: 123,
 *       id: 123,
 *       width: 123,
 *       position: 123,
 *       type: 'asdf',
 *       title: 'asdf'
 *    }],
 *  }]
 * }
 */
router.get('/', async (req, res) => {
  console.debug('GET request to dashboard');
  const dashboards = await Dashboard.findAll({
    include: [
      {
        model: DashboardGraph,
        include: [DashboardSource],
      },
    ],
  });

  res.send(dashboards);
});

/**
 * POST
 * Creates a new dashboard
 *
 * params: {
 *   name: 'station name'
 * }
 *
 * returns: 200 success, 400 if validation fails
 */
router.post('/', async (req, res) => {
  console.debug('POST request to dashboard');
  const { name } = req.body;

  const valid = validateDashboardParams(req.body);
  if (valid.error) {
    res.status(400).send({ error: valid.error });
    return;
  }

  try {
    Dashboard.create({ name });
    console.debug('new dashboard created.');
    res.status(200).send({ message: 'success' });
  } catch (error) {
    console.error('an error occured!');
    console.error(error);
    res.status(400).send({ error: 'An unknown error has occured' });
  }
});

export default router;
