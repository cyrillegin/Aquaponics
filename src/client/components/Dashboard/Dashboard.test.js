import { mount } from 'enzyme';
import React from 'react';
import Dashboard from './Dashboard';

describe('dashboard', () => {
  it('should render a snap shot', () => {
    const wrapper = mount(<Dashboard />);
    expect(wrapper).toMatchSnapshot();
  });
});
