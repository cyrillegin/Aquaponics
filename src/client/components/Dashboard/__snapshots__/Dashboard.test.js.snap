// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dashboard should render a snap shot 1`] = `"<div class=\\"css-c8dous \\">0Loading...</div>"`;
